### PR TITLE
PHP 7.2 compatibility and adding example usage

### LINKED ISSUE #10: Not php7.2 compatible
Got the " Function mcrypt_encrypt() is deprecated" error message, same thing like using:

pecl install mcrypt-1.0.1

What a nightmare.

### LINKED ISSUE #11: Documentation is missing
There is no example about how to use this API. It should be great to include at least some line of code into the README or an other file..

### PR DESCRIPTION
**Pull request by jstruzik**
This PR resolves https://github.com/robinpowered/php-ntlm/issues/10 and resolves https://github.com/robinpowered/php-ntlm/issues/11.

We now require PHP 7.1 and will mention this BC break in the latest tagged release. Main reasoning behind this is PHP's introduction of the [`HashContext` class](https://www.php.net/manual/en/class.hashcontext.php) in 7.2 since an `instanceof` check would require us to still attempt to import the class for earlier versions of PHP.

I've updated some typing along the way as well but this whole library could use an update.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -17,6 +14,7 @@ matrix:
 
 before_install:
    - export ALLOW_FAILURE=1; if [ "$TRAVIS_PHP_VERSION" != "hhvm" ] && [ "$TRAVIS_PHP_VERSION" != "nightly" ]; then export ALLOW_FAILURE=0; fi
+   - pecl install mcrypt-1.0.2
```
Comment (Rican7) [inline .travis.yml]: You're going to need `yes '' | pecl install mcrypt-1.0.2` for this to work on Travis-CI, to accept the default values in the compilation prompts.

**Comment by Rican7** [comment]
Awesome work on this!

> We now require PHP 7.2 and will mention this BC break in the latest tagged release. Main reasoning behind this is PHP's introduction of the [`HashContext` class](https://www.php.net/manual/en/class.hashcontext.php) in 7.2 since an `instanceof` check would require us to still attempt to import the class for earlier versions of PHP.

I'm not sure I'm following this. PHP `use` statements aren't imports, they're just "aliases" basically, [so using a `use` statement for a class that isn't available won't cause any harm or error, and `instanceof` is designed to fail gracefully (a boolean `false`) even if the type that you use as an "argument" (its a construct symbol) doesn't exist](https://3v4l.org/kjLLs).

All that being said, updating to a PHP 7.2 requirement isn't bad, but it may be unnecessary. It would make sense, though, to require 7.1, as its the [minimum supported version of the language itself as of now](https://www.php.net/supported-versions.php), and it would allow the added scalar typing. 😃 

> I've updated some typing along the way as well but this whole library could use an update.

Ha, true.

File: src/Robin/Ntlm/Crypt/Hasher/AbstractHasher.php [hunk 2]
```diff
@@ -90,14 +94,14 @@ public function digest()
...
      * @link http://php.net/manual/en/hash.resources.php
      * @param mixed $context The context to validate.
-     * @return resource The incremental hashing context.
+     * @return resource|HashContext The incremental hashing context.
      * @throws InvalidArgumentException If the hash context isn't valid.
      */
     protected function validateHashContext($context)
     {
-        if (false === $context
+        if (!($context instanceof HashContext) && (false === $context
```
Comment (vnwhlr) [inline src/Robin/Ntlm/Crypt/Hasher/AbstractHasher.php:102]: Is the `HashContext` class 100% interchangeable with the `Hash Context` resource type?

**Review by vnwhlr — APPROVED** [review]

File: src/Robin/Ntlm/Crypt/Hasher/AbstractHasher.php (hunk 2 above)
Comment (Rican7) [reply src/Robin/Ntlm/Crypt/Hasher/AbstractHasher.php:102]: It _should_ be (I believe), as its just used as an argument to functions to keep an internal context and has no methods.

https://www.php.net/manual/en/class.hashcontext.php
https://www.php.net/manual/en/function.hash-init.php#refsect1-function.hash-init-changelog

**Review by Rican7 — APPROVED** [review]
Thanks Jake!

We should probably update the description to match the new changes. Otherwise, this looks great! 👍

**Pull request merged**